## Association Rule EBC

### Problem Statement:
 * What are the education level that may be frequently give feedback?

### Objective:
* To learn how Apriori Algorithm and Association Rules works.
* To learn how Combination and Permutation helps to find Support and Confidence of itemsets respectively.
* To find frequent itemsets with high confidence and lift, keeping both item together will help to increase sales.


### Introduction
* Association rule mining is one of an important technique of data mining for knowledge discovery.
* The knowledge of the correlation between the items in the data transaction can use association rule mining.
* Retail store analysis is one of an application area of association rule mining technique.
* The possible percentage of the correlation of combined items gives the new knowledge. Therefore, it is a very helpful for determiner to take the decisions

### Analysis

In [1]:
# Importing Required Library

import pandas as pd
import numpy as np

In [2]:
DATA_FILE_PATH = 'C:\\Users\\Yostina\\SentimentNLP\\'
DATA_FILE_NAME = 'sentiment_mapped_datas4.csv'
SAVE_FILE = True

In [3]:
# Reading Excel file 
df = pd.read_csv(DATA_FILE_PATH+DATA_FILE_NAME,encoding="utf-8" )

In [4]:
df

Region Sex Age Edu_Level  \
0      Debub   M  50      Cert   
1      Debub   M  50      Cert   
2      Debub   M  50      Cert   
3      Debub   M  50      Cert   
4      Debub   M  30   Sec. Sc   
...      ...  ..  ..       ...   
3844  Amhara   M  48   Sec. Sc   
3845  Amhara   M  48   Sec. Sc   
3846  Amhara   M  48   Sec. Sc   
3847  Amhara   M  48   Sec. Sc   
3848  Amhara   M  48   Sec. Sc   

                                               Feedback  Fb_Day  Fb_Month  \
0              ኢቲቪ ሰራተኞች በሙሉ እንኳን ለገና በአል በሰላም አደረሳችሁ፡፡       1         1   
1     መንግስት በአገራችን በደረሰው የተለያዩ ችግሮች ላይ እየወሰደ ያለው እርም...       1         1   
2     ኢቲቪ ደግሞ በዜና በወቅታዊ ጉዳዮች ላይ እውነት ለመናገር በጣም ጥሩ እየ...       1         1   
3     ኢቲቪ ቢሰራ የምንለው ህዝቡ ስሩን የሰደደው የህዋት አስተዳደር አሁንም ባ...       1         1   
4     ኢቲቪ ዜና 57 ላይ የሚቀርቡት ወቅታዊ ዜናዎች በሀገርና በውጭ በጣም ጥሩ...       1         1   
...                                                 ...     ...       ...   
3844  ኢ.ሬ. ከ2011 ጥር ጀምሮ ጥራት የለው ስርጭት እየተገለገልን አይደለም ...      30        12   
3845  ኢ.ሬ. 180 ፕሮግራም ከ3-6፡00 የሚቀርበው የተሰጦ ውድድር አንዳንዴ ...      30        12   
3846  ኤዔፍ ኤዔም 97.1 Debube ውስጥ አይሰማም ኢቢሲ በዚህ ላይ ቢሰራና ...      30        12   
3847  የጥያቄና መልስ በአዲስ ዘመን በኢትዮጵያዊነትና በብሄር ብሄረሰቦች ላይ የ...      30        12   
3848        ኢ.ሬ. ስርጭት ለገዳን ወረዳ ላይ በየግዜው ይቆራረጣል ቢስተካከል፡፡      30        13   

      Fb_Year                                             tokens category  \
0      2008.0  ['ኢቲቪ', 'ሰራተኞች', 'በሙሉ', 'እንኳን', 'ለገና', 'በአል', ...      ETV   
1      2008.0  ['መንግስት', 'በአገራችን', 'በደረሰው', 'የተለያዩ', 'ችግሮች', ...      ETV   
2      2008.0  ['ኢቲቪ', 'ደግሞ', 'በዜና', 'በወቅታዊ', 'ጉዳዮች', 'ላይ', '...      ETV   
3      2008.0  ['ኢቲቪ', 'ቢሰራ', 'የምንለው', 'ህዝቡ', 'ስሩን', 'የሰደደው',...      ETV   
4      2008.0  ['ኢቲቪ', 'ዜና', '57', 'ላይ', 'የሚቀርቡት', 'ወቅታዊ', 'ዜ...      ETV   
...       ...                                                ...      ...   
3844   2013.0  ['ኢ', '.', 'ሬ', '.', 'ከ2011', 'ጥር', 'ጀምሮ', 'ጥራ...      ETV   
3845   2013.0  ['ኢ', '.', 'ሬ', '.', '180', 'ፕሮግራም', 'ከ3', '-'...      ETV   
3846   2013.0  ['ኤዔፍ', 'ኤዔም', '97.1', 'Debube', 'ውስጥ', 'አይሰማም...      ETV   
3847   2013.0  ['የጥያቄና', 'መልስ', 'በአዲስ', 'ዘመን', 'በኢትዮጵያዊነትና', ...      ETV   
3848   2013.0  ['ኢ', '.', 'ሬ', '.', 'ስርጭት', 'ለገዳን', 'ወረዳ', 'ላ...      ETV   

      sentiment  
0             0  
1             1  
2             1  
3             0  
4             1  
...         ...  
3844          0  
3845          0  
3846          1  
3847          0  
3848          0  

[3849 rows x 11 columns]

In [5]:
## dropping Duplicate feedback
df = df.drop_duplicates()

In [6]:
new = df[['Region','Sex','category','Edu_Level']]

In [7]:
import warnings
warnings.filterwarnings('ignore')

In [8]:
## assigning column to data frame "df"
df['Region'] = new['Region']
df['Sex'] = new['Sex']
df['category'] = new['category']
df['Edu_Level'] = new['Edu_Level']

In [9]:
# in this dataframe we only need column category and Edu_Level, rest is not needed in association mining rule
df[['Region', 'Sex', 'category', 'Edu_Level']].head(10)
                                                    

Region Sex category Edu_Level
0  Debub   M      ETV      Cert
1  Debub   M      ETV      Cert
2  Debub   M      ETV      Cert
3  Debub   M      ETV      Cert
4  Debub   M      ETV   Sec. Sc
5  Debub   M      ETV   Sec. Sc
6  Debub   M      ETV   Sec. Sc
7  Debub   M      ETV   Sec. Sc
8  Debub   M      ETV       Dip
9  Debub   F      ETV       Ele

In [10]:
# we need to convert cloumn category & Edu_Level into Crosstab or we can say Binary Matrix as follow
media = pd.crosstab(index= df['category'], columns= df['Edu_Level'])
media

Edu_Level  Cert  Degree  Dip  Dip   Ele  Ilitrate  Msc  Sec. Sc  dip  doctor
category                                                                    
ETV         154     695  385     2  989         6   10     1163  335       1
FM            0       1    0     0    4         0    0        1    0       0
NR            0       4    2     0   10         0    0       15   11       0

In [11]:
media

Edu_Level  Cert  Degree  Dip  Dip   Ele  Ilitrate  Msc  Sec. Sc  dip  doctor
category                                                                    
ETV         154     695  385     2  989         6   10     1163  335       1
FM            0       1    0     0    4         0    0        1    0       0
NR            0       4    2     0   10         0    0       15   11       0

### Creating APRIORI_MY function to generate frequent itesets based on minimum threshold support = 0.02

In [12]:
def APRIORI_MY(data, min_support=0.04,  max_length = 4):
    # Collecting Required Library
    import numpy as np
    import pandas as pd
    from itertools import combinations
    # Step 1:
    # Creating a dictionary to stored support of an itemset.
    support = {} 
    L = list(data.columns)
    
    # Step 2: 
    #generating combination of items with len i in ith iteration
    for i in range(1, max_length+1):
        c = set(combinations(L,i))
        
    # Reset "L" for next ith iteration
        L =set()     
    # Step 3: 
        #iterate through each item in "c"
        for j in list(c):
            #print(j)
            sup = data.loc[:,j].product(axis=1).sum()/len(data.index)
            if sup > min_support:
                #print(sup, j)
                support[j] = sup
                
                # Appending frequent itemset in list "L", already reset list "L" 
                L = list(set(L) | set(j))
        
    # Step 4: data frame with cols "items", 'support'
    result = pd.DataFrame(list(support.items()), columns = ["Items", "Support"])
    return(result)

In [13]:
## finding frequent itemset with min support = 4%
my_freq_itemset = APRIORI_MY(media, 0.04, 3)
my_freq_itemset.sort_values(by = 'Support', ascending = False)

Items       Support
107  (Sec. Sc, Degree, Ele)  2.664648e+08
120     (Sec. Sc, Ele, Dip)  1.476100e+08
63      (Sec. Sc, dip, Ele)  1.284403e+08
162  (Sec. Sc, Degree, Dip)  1.037299e+08
71   (Sec. Sc, Degree, dip)  9.025871e+07
..                      ...           ...
29       (Ilitrate, doctor)  2.000000e+00
4               (Ilitrate,)  2.000000e+00
46           (Dip , doctor)  6.666667e-01
5                   (Dip ,)  6.666667e-01
2                 (doctor,)  3.333333e-01

[175 rows x 2 columns]

### Creating ASSOCIATION_RULE_MY function to generate itemset based on minimun threshold confidence.

In [21]:
def ASSOCIATION_RULE_MY(df, min_threshold=0.5):
    import pandas as pd
    from itertools import permutations
    
    # STEP 1:
    #creating required varaible
    support = pd.Series(df.Support.values, index=df.Items).to_dict()
    data = []
    L= df.Items.values
    
    # Step 2:
    #generating rule using permutation
    p = list(permutations(L, 2))
    
    # Iterating through each rule
    for i in p:
        
        # If LHS(Antecedent) of rule is subset of RHS then valid rule.
        if set(i[0]).issubset(i[1]):
            conf = support[i[1]]/support[i[0]]
            #print(i, conf)
            if conf > min_threshold:
                #print(i, conf)
                j = i[1][not i[1].index(i[0][0])]
                lift = support[i[1]]/(support[i[0]]* support[(j,)])
                leverage = support[i[1]] - (support[i[0]]* support[(j,)])
                convection = (1 - support[(j,)])/(1- conf)
                data.append([i[0], (j,), support[i[0]], support[(j,)], support[i[1]], conf, lift, leverage, convection])

         
    # STEP 3:
    result = pd.DataFrame(data, columns = ["antecedents", "consequents", "antecedent support", "consequent support",
                                        "support", "confidence", "Lift", "Leverage", "Convection"])
    return(result)

In [23]:
## Rule with minimun confidence = 50%
my_rule = ASSOCIATION_RULE_MY(my_freq_itemset, 0.5)
my_rule

ZeroDivisionError: float division by zero

### Finally sorting results by Lift to get highly associated itemsets.

In [16]:
my_rule.sort_values(by='Lift', ascending= False).head(10)

NameError: name 'my_rule' is not defined

## Cross Verifying results with  apriori and association rule from mlxtend -> frequent patterns package

In [17]:
# Loading standard package
from mlxtend.frequent_patterns import apriori, association_rules

In [18]:
## finding frequent itemset with min support = 4%
frequent_itemset = apriori(df = media, min_support= 0.04, use_colnames= True)
frequent_itemset.sort_values(by = 'support', ascending = False)

ValueError: The allowed values for a DataFrame are True, False, 0, 1. Found value 154

### Createing associate rule such that item brought with conditional probability(Confidence) more than 50% with corresponding item

In [19]:
## Rule with minimun confidence = 50%
Rules = association_rules(frequent_itemset, min_threshold= 0.5)
Rules

NameError: name 'frequent_itemset' is not defined

In [20]:
## Finally sorting results by Lift to get highly associated itemsets.
Rules.sort_values(by='lift', ascending= False).head(10)

NameError: name 'Rules' is not defined

### Conclusion
 * Results from developed function(APRIORI_MY, ASSOCIATION_RULE_MY) has matched with builts packages.
 * it is observed that "Toast" & "Coffee" are highly associated with lift 1.48.
 * Coffee has been brought most frequently with 47.5% of all the transcaction